File: Subliminal.Tests.CSharp/TestLog.cs
using FluentAssertions;
using System.Collections.Generic;
using Xunit;

namespace Subliminal.Tests.CSharp
{
    public class TestLog
    {
        [Fact]
        public void EmitsDataWhenLogged()
        {
            var log = new Log<object>();
            var subscriber = new Queue<object>();
            log.Subscribe(subscriber.Enqueue);
            var data1 = new object();
            var data2 = new object();
            var data3 = new object();
            log.Log(data1);
            log.Log(data2);
            log.Log(data3);
            subscriber.ToArray().Should().Equal(data1, data2, data3);
        }

        [Fact]
        public void SubscribersReceiveTheSameData()
        {
            var log = new Log<object>();
            var subscriber1 = new Queue<object>();
            var subscriber2 = new Queue<object>();
            log.Subscribe(subscriber1.Enqueue);
            log.Subscribe(subscriber2.Enqueue);
            var data1 = new object();
            var data2 = new object();
            var data3 = new object();
            log.Log(data1);
            log.Log(data2);
            log.Log(data3);
            subscriber1.ToArray().Should().Equal(data1, data2, data3);
            subscriber2.ToArray().Should().Equal(data1, data2, data3);
        }
    }
}

File: Subliminal.Tests.FSharp/TestLog.fs
namespace Subliminal.Tests

open Subliminal
open Swensen.Unquote
open System.Collections.Generic
open Xunit

module ``Test Log<'Data>`` =
    [<Fact>]
    let ``emits data when logged`` () =
        let log = Log<obj>()
        let subscriber = Queue()
        log |> Log.subscribeForever subscriber.Enqueue
        let data1 = obj()
        let data2 = obj()
        let data3 = obj()
        log.Log(data1)
        log.Log(data2)
        log.Log(data3)
        test <@ subscriber.ToArray() = [| data1; data2; data3 |] @>

    [<Fact>]
    let ``subscribers receive the same data`` () =
        let log = Log<obj>()
        let subscriber1 = Queue()
        let subscriber2 = Queue()
        log |> Log.subscribeForever subscriber1.Enqueue
        log |> Log.subscribeForever subscriber2.Enqueue
        let data1 = obj()
        let data2 = obj()
        let data3 = obj()
        log.Log(data1)
        log.Log(data2)
        log.Log(data3)
        test <@ subscriber1.ToArray() = [| data1; data2; data3 |] @>
        test <@ subscriber2.ToArray() = [| data1; data2; data3 |] @>
